stdlib/camlinternalLazy.ml

```diff
@@ -59,83 +54,73 @@ external reset_to_lazy : Obj.t -> unit = "caml_lazy_reset_to_lazy" [@@noalloc]
-    end;
+    Obj.set_field b 0 (Obj.repr (fun () -> raise e));
     reset_to_lazy b;
     raise e
 
 (* Assumes [blk] is a block with tag forcing *)
 let do_force_val_block blk =
   let b = Obj.repr blk in
   let closure = (Obj.obj (Obj.field b 0) : unit -> 'arg) in
-  Obj.set_field b 0 (domain_unique_token ());
+  Obj.set_field b 0 (Obj.repr ()); (* Release the closure *)
   let result = closure () in
   Obj.set_field b 0 (Obj.repr result);
   update_to_forward b;
   result
 
+(* Called by [force_gen] *)
 let force_gen_lazy_block ~only_val (blk : 'arg lazy_t) =
+  (* In the common case, expect the tag to be [lazy_tag], but may be other tags
+     due to concurrent forcing of lazy values. *)
   match update_to_forcing (Obj.repr blk) with
   | 0 when only_val -> do_force_val_block blk
   | 0 -> do_force_block blk
-  | 1 -> raise Undefined
   | _ ->
-      (* We raise [RacyLazy] for all the other cases.
-       *
-       * If the return value was 2, we should raise RacyLazy. We had observed
-       * earlier in [force_gen] that the [blk] was not [Obj.forward_tag] and
-       * not short-circuited. But now we observe that the value has either been
-       * forwarded (return value 3) or forwarded and short-circuited (4). Both
-       * of these indicate a race. *)
-      raise RacyLazy
+      (* If the return value is 1, then the tag is [forcing_tag]. We are either
+         recursively forcing on the same domain, or concurrently forcing from
+         multiple domains. Raise [Undefined].
+
+         If the return value is 2, then the tag is [forward_tag]. In
+         [force_gen], we checked whether the tag was [forward_tag]. It has
+         changed since the last read. Hence, there is another domain which is
+         concurrently forcing this lazy. Raise [Undefined].
+
+         If the return value is 3, then the tag is not [forcing_tag] or
+         [forward_tag], which we checked in [force_gen]. The tag is not
+         [lazy_tag] or otherwise [update_to_forcing] would have succeeded and
+         returned [0]. The tag has changed from the expected [lazy_tag] due to
+         another domain concurrently, and the GC short circuiting the result.
+         Raise [Undefined].
+         *)
+      raise Undefined
 
 (* used in the %lazy_force primitive *)
 let force_lazy_block blk = force_gen_lazy_block ~only_val:false blk
 
-let try_force_gen_lazy_block ~only_val (blk : 'arg lazy_t) =
-  match update_to_forcing (Obj.repr blk) with
-  | 0 when only_val -> Some (do_force_val_block blk)
-  | 0 -> Some (do_force_block blk)
-  | 1 -> raise Undefined
-  | _ ->
-      (* Other cases indicate racy access to lazy for the same reasons outlined
-       * in [force_gen_lazy_block]. *)
-      None
-
 (* [force_gen ~only_val:false] is not used, since [Lazy.force] is
    declared as a primitive whose code inlines the tag tests of its
    argument, except when afl instrumentation is turned on. *)
-
 let force_gen ~only_val (lzv : 'arg lazy_t) =
```

Comment:
```suggestion
let force_gen ~only_val (lzv : 'arg lazy_t) =
  let lzv = Sys.opaque_identity lzv in
```
The above fix was introduced in trunk to prevent flambda from forwarding the tag information from a previous read (which can become invalid with short-circuiting). (See https://github.com/ocaml/ocaml/issues/7301 and https://github.com/ocaml/ocaml/issues/713.) Do you need something similar? Are there risks that similar issues happen with the new tag Forcing_tag? (Note: this can be addressed in a separate PR if needed.)

Reply:
I'll merge this suggestion. And I think the rest of the discussion may happen at ocaml/ocaml after the MVP merge to trunk, before the 5.00 release. The MVP will not have flambda support. If the discussion and the PR happens after merge, we will have more eyes on the flambda support as a whole.

Reply:
Let's remind ourselves to keep the original comment describing the problem when merging upstream, and to go back to it later.